### PR TITLE
Grade abbreviation syncing

### PR DESCRIPTION
**Pull request by DorisWongNHS**
- Create `grade` table in Sync
- Enrich trainee placement grade abbreviation according to the grade table
- Trigger placement re-enrichment after grade is updated

TIS21-3123

### CODE REVIEW

File: src/main/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacade.java [hunk 1]
```diff
@@ -72,10 +78,12 @@ public class PlacementEnricherFacade {
 
   PlacementEnricherFacade(PostSyncService postService,
       TrustSyncService trustService, SiteSyncService siteService,
+      GradeSyncService gradeServie,
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacade.java:81]: Just for consistency, could we call this gradeService

File: build.gradle [hunk 2]
```diff
@@ -10,7 +10,7 @@ plugins {
 }
 
 group = "uk.nhs.hee.tis.trainee"
-version = "0.31.0"
+version = "0.31.1"
```
Comment (ReubenRobertsHEE) [inline build.gradle:13]: I'd be tempted to call this a minor version update, not a bugfix :)

File: src/main/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacade.java (hunk 1 above)
Comment (DorisWongNHS) [reply src/main/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacade.java]: Oh missing typo. Thanks for the good spot.

**Comment by ReubenRobertsHEE** [comment]
This is more of an open question: do we need to handle the case where a grade changes, i.e. if it's abbreviation is updated? I don't know if that ever happens (@Judge40 not sure if you do, or whether we should check with James or others)

For other things like Programmes etc. we have an event listener that triggers re-enrichment of related objects after a programme is saved, to cascade that update into the related records. If grades records can change I think we'd need something similar here.

**Comment by ReubenRobertsHEE** [comment]
Any chance of bumping up the test coverage a bit for the PlacementEnricherFacade changes? These might be quick to write  tests for:
.enrich(Placement placement, Grade grade) 
.enrichPlacementWithRelatedGrade(Placement placement)
.populateGradeDetails(Placement placement, String gradeAbbr)

**Review by ReubenRobertsHEE — CHANGES_REQUESTED** [review]
I think we do need the event listener for the grade being saved I'm afraid 😁 . 

Consider this scenario: a placement record arrives with a gradeId for a grade that has not yet been saved. The existing code correctly fails to find the grade abbreviation, and does not sync the placement. It also makes a request for the grade from TCS, which is good.

But when the requested grade record arrives from TCS, there is no re-enriching of the related placements and syncing of them. So the trainee whose original placement referred to a grade that didn't exist in the database at that point would simply never get their placement added to their profile afaik.

Since we can't be sure about the order of stuff coming from the DMS, I think we can't assume that all the grade record messages will be received before placement messages.

What do you guys think?

**Comment by ReubenRobertsHEE** [comment]
> 

Thanks Doris - when I was testing locally I must have pulled the code before you made that commit. Let me take a look now...

**Review by ReubenRobertsHEE — DISMISSED** [review]
LGTM, nice work guys.

One question: given we will have a bunch of existing trainees in TSS with profile placements having the wrong grade abbreviation, are you thinking to create another subtask on this ticket to correct this data?

**Comment by DorisWongNHS** [comment]
@ReubenRobertsHEE, yeah there was a plan to reload table data of reference.Grade from DMS then check the trainee profiles as part of the verify on Preprod/Prod process. Do you think there would be extra step that we need to do?

**Comment by ReubenRobertsHEE** [comment]
> @ReubenRobertsHEE, yeah there was a plan to reload table data of reference.Grade from DMS then check the trainee profiles as part of the verify on Preprod/Prod process. Do you think there would be extra step that we need to do?

Yeah, I think that should work. The event handler for saving each record should re-enrich the related placements, so they'll all get a refresh. I think there are actually about ~34 grades where there is bad data (out of ~500 in total) (as per `SELECT distinct g.* FROM tcs.Placement p JOIN reference.Grade g ON p.gradeId = g.id WHERE p.gradeAbbreviation <> g.abbreviation` ) but it is probably easiest to just reload the entire table! 😀

**Review by ReubenRobertsHEE — APPROVED** [review]

**Pull request merged**